feat(toast)!: migrate toast to core tokens

yosevu opened the pull request:
## Description

This PR migrates the Toast component to core-tokens.

- Removes deprecated variants (warning, error, success)

## How and where has this been tested?

- Tested locally referencing the Components batch 2 XD file.
- Tested with VoiceOver on Safari
- Tested WHCM with forced colors in Chrome Emulation

## Screenshots
<!-- If applicable, add screenshots to show what you changed -->

## To-do list
~~- [ ] If my change impacts other components, I have tested to make sure they don't break.~~
~~- [ ] If my change impacts documentation, I have updated the documentation accordingly.~~
- [x] I have read the [CONTRIBUTING document](/.github/CONTRIBUTING.md).
- [x] I have tested these changes in Windows High Contrast mode.
<!-- If this pull request isn't ready, add any remaining tasks here -->
- [x] This pull request is ready to merge.

yosevu commented on components/toast/index.css:
```diff
@@ -86,12 +184,15 @@ governing permissions and limitations under the License.
 
   .spectrum-Button {
+    /* https://caniuse.com/mdn-css_properties_float_flow_relative_values */
     float: inline-end;
```
This only works on Firefox.  Looking for a better supported approach.

yosevu replied:
`float: inline-end` and `:dir(ltr) { float: right; }` are the only approaches I found, but both are Firefox only. Should I have `float: right` as a fallback even though it won't work for RTL direction @pfulton?

bernhard-adobe replied:
Can you use a browser-non specific version?

yosevu replied:
Hey Bernhard, both options that I have found are only ones I found that are based on directionality. `float: right;` is used currently. I haven't found another approach.

bernhard-adobe replied:
Awesome! thanks for checking! `float: right` it is!

yosevu commented on components/toast/index.css:
```diff
@@ -30,54 +79,102 @@ governing permissions and limitations under the License.
+  &:lang(js),
+  &:lang(zh),
+  &:lang(ko) {
+    line-height: var(--spectrum-toast-line-height-CJK, var(--mod-toast-line-height, var(--spectrum-toast-line-height)));
```
I changed the token  naming format `--system-component-style-description-modifier` to try to be consistent. Are there any issues with this approach?

pfulton replied:
I think this is fine. Thanks for the heads up!

bernhard-adobe commented on components/toast/index.css:
```diff
@@ -10,18 +10,60 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 governing permissions and limitations under the License.
 */
 
+/* toast/index.css
```
Maybe remove this code if we don't need it anymore

bernhard-adobe commented on components/toast/index.css:
```diff
@@ -10,18 +10,60 @@ OF ANY KIND, either express or implied. See the License for the specific languag
-    --spectrum-global-dimension-size-130
-  );
+  /* Hardcoded variables */
+  --spectrum-toast-font-weight: 400;
```
maybe you could use this custom-var for now:
https://github.com/adobe/spectrum-css/blob/main/components/tokens/custom-spectrum/custom-vars.css#L32
`  --spectrum-font-weight-regular: 400;`

bernhard-adobe commented on components/toast/index.css:
```diff
@@ -10,18 +10,60 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+
+  /* Size */
+
+  --spectrum-toast-font-size: var(--spectrum-font-size-100);
```
I think in all of these definitions we need to add the `--mod`  option as well to enable custom overwrites. 

like `  --spectrum-toast-font-size: var(--mod-font-size-100, var(--spectrum-font-size-100);
`

more details here: https://github.com/adobe/spectrum-css/wiki/Migrating-a-component-to-core-tokens#implementing-the-custom-properties-api

yosevu replied:
Oh I added the `--mod` prefix when I use the component token e.g. `font-size: var(--mod-toast-font-size, var(--spectrum-toast-font-size));`. Is that what you mean?

bernhard-adobe replied:
yes, just checking this is consistently used.

bernhard-adobe commented on components/toast/index.css:
```diff
@@ -10,18 +10,60 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+  --spectrum-toast-corner-radius: var(--spectrum-corner-radius-100);
+
+  --spectrum-toast-line-height: var(--spectrum-line-height-100);
+  --spectrum-toast-line-height-CJK: var(--spectrum-CJK-line-height-100);
```
We may want to use for CJK support the lang-selector:
```CSS
 &:lang(ja),
  &:lang(zh),
  &:lang(ko) {
    line-height: var(--mod-toast-line-height-CJK, var(----spectrum-toast-line-height-CJK));
  }

yosevu replied:
I used the lang selector below in the `.spectrum-Toast-content` styles. Is that what you meant?

bernhard-adobe replied:
maybe we should used that when referencing CJK line-heights? @pfulton what do you suggest?

pfulton replied:
@bernhard-adobe could you clarify? I think @yosevu implemented it as I'd expect [here](https://github.com/adobe/spectrum-css/blob/3e3807d310b77681cd07660e10bbd970aedfcfdd/components/toast/index.css#L138-L142).

@yosevu You may have a typo on line 138: `&:lang(js)` should be `&:lang(ja)`, right?

Also, can we lowercase `CJK` in the custom property name, please?

bernhard-adobe replied:
@pfulton could we lower-case CJK in core tokens name, please ;)?

yosevu replied:
Ah thanks. Updated in e3f4a568096fb55054c773f48d783c9a00994d2b

bernhard-adobe commented on components/toast/index.css:
```diff
@@ -10,18 +10,60 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+
+@media (forced-colors: active) {
+  .spectrum-Toast {
+    border: 1px solid transparent;
```
Maybe we could soft-code this `1px` value? A bit of a nit, not a bit priority or blocker

pfulton replied:
Can we use the custom property defined [here](https://github.com/adobe/spectrum-css/pull/1483/files#diff-dd286afa975e8faf641af07455647fd4d891bcc2997ac531249fa23a6dad2b07R65)?

```css
border: var(--mod-toast-border-width, var(--spectrum-toast-border-width)) solid transparent;
```

yosevu replied:
Yes. I think this is outdated. I refactored this to use that property.

bernhard-adobe commented on components/toast/index.css:
```diff
@@ -30,68 +72,120 @@ governing permissions and limitations under the License.
+  color: var(--mod-toast-background-color-default, var(--spectrum-toast-background-color-default));
+
+  &--negative,
+    /* @deprecated */.spectrum-Toast--error {
```
remove if deprecated?

pfulton replied:
This is a great question that we should maybe surface to design and/or Jess. I'm inclined to think that we should only show current guidance and get rid of the "use X variant instead" examples in the docs.

pfulton replied:
Asked [here](https://adobedesign.slack.com/archives/G019JTYMT6H/p1659387331617259)

yosevu replied:
I removed the deprecated variants.

bernhard-adobe commented on components/toast/index.css:
```diff
@@ -30,68 +72,120 @@ governing permissions and limitations under the License.
+  }
+
+  &--positive,
+    /* @deprecated */.spectrum-Toast--success {
```
remove if deprecated?

bernhard-adobe reviewed (changes requested):
a couple things that we may want to work on. Mostly the `--mod` custom modifier that seems to e missing from a couple of values.

yosevu commented on components/toast/index.css:
```diff
@@ -10,18 +10,60 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+
+@media (forced-colors: active) {
+  .spectrum-Toast {
+    border: 1px solid transparent;
```
Updated.

yosevu commented on components/toast/index.css:
```diff
@@ -30,68 +72,120 @@ governing permissions and limitations under the License.
+  margin-block-start: var(--mod-toast-spacing-top-edge-to-divider, var(--spectrum-toast-spacing-top-edge-to-divider));
+  margin-block-end: var(--mod-toast-spacing-bottom-edge-to-divider, var(--spectrum-toast-spacing-bottom-edge-to-divider));
+
+  border-inline-start-color: rgba(255, 255, 255, 0.2);
```
Yeah, makes sense to use it base don that note. Thank you.

pfulton commented:
Beta released:
`- @spectrum-css/toast@8.0.0-beta.0`

pfulton commented:
New beta released due to faulty build:
`- @spectrum-css/toast@8.0.0-beta.1`

pfulton commented:
@yosevu We're seeing some size changes and close button issues when we try to integrate this with the Spectrum Web Components project.

[Here's my draft PR](https://github.com/adobe/spectrum-web-components/pull/2485) where you can take a look at their VRT results.

Would you mind double-checking on the sizing and maybe also try bringing in the latest release of the CloseButton? That might actually help things.

bernhard-adobe reviewed (commented):
Approved but won't hold clicking the approval button.  Still need to address however the comment by @pfulton about the impact on Spectrum Web Components

yosevu commented:
I updated the token used for the `padding-inline` of the close button. I believe this was the issue based on the comment in @pfulton draft PR. Would you mind taking a look confirm it @bernhard-adobe? I also updated to the latest CloseButton package.

Design:
<img width="1213" alt="Screen Shot 2022-08-10 at 10 14 35 PM" src="https://user-images.githubusercontent.com/16216104/183910581-4fd64bd2-f895-4ffc-9add-ba52e6fbedd8.png">
Beta:
<img width="796" alt="Screen Shot 2022-08-10 at 10 16 48 PM" src="https://user-images.githubusercontent.com/16216104/183911293-076feb8e-d685-4375-9844-bc89a7af8558.png">

Updated:
<img width="773" alt="Screen Shot 2022-08-10 at 10 17 39 PM" src="https://user-images.githubusercontent.com/16216104/183911287-42294c66-4937-487a-b8b5-22b8d15c9b86.png">

bernhard-adobe commented:
VRT is: https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/31/

bernhard-adobe commented:
VRT test is 	https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/31/artifact/backstop_data/html_report/index.html
[2022-08-10 -- 1483 - toast.zip](https://github.com/adobe/spectrum-css/files/9306260/2022-08-10.--.1483.-.toast.zip)

pfulton commented:
Released:
`8.0.0-beta.2`

bernhard-adobe reviewed (approved):
released as beta so I would approve this PR ;)

yosevu commented:
Could I get a new release for this @pfulton  or @GarthDB?

pfulton commented:
Released:
`@spectrum-css/toast@8.0.0-beta.3`

pfulton commented:
I think we're ready to merge this in and graduate the release.
